src/toil/test/wdl/builtinTest.py

```diff
@@ -341,6 +341,7 @@ def setUpClass(cls):
         super(WdlLanguageSpecWorkflowsTest, cls).setUpClass()
         cls.test_path = os.path.abspath("src/toil/test/wdl/wdl_specification")
 
+    @unittest.skip('WIP')
```

Comment:
Unfortunately, lazy-evaluation doesn't quite work when the return value of a function call is used immediately (e.g.: `write_json(read_lines(test_file))`). 

My current thought is to make `WDLFunctionCall` act like the evaluated object by substituting `self._resolve()` in the special methods (e.g.: `__str__`, `__getattr__`), but that might introduce more unintended problems.. I may have to rethink my approach, though, I think we need some sort of lazy-evaluation to deal with Toil Promises.

Reply:
We also need to lazy-chain functions, right? If you have a promise `p`, and the user says `func1(p)`, you need to defer evaluation of `func1` until `p` is filled in. But if you say `func2(func1(p))`, you can't evaluate `func1(p)` right now just because `func2` wants to access part of it; the whole thing needs to be lazy. But also, `func2` expects the result of `func1`, not a `WDLFunctionCall`...

Maybe resolving on `__getattr__` will Just Work?

I'm concerned we can't/shouldn't try to be lazy across WDL tasks; if you write a file in one task and pass it to a bunch of scattered tasks, you don't expect the write to re-happen in all those scattered tasks. You want it to have happened once in the initial scattering task, even though the file might not make it into a command line until the scattered tasks try to use it.

But at the same time, we want a task to be able to return the result of its child tasks, subject to some WDL post-processing, which can't happen while the task itself is running.

So it sounds like every WDL task might really need to be two jobs: we need a follow-on job (or a final child whose result is returned) to run the WDL code that depends on the child jobs' results.

Reply:
Maybe since WDL code is very limited we can just force it to be evaluated in whatever job actually wants to pass its result to a command, and drag all the input data along in closures. It can be pass-by-comnpletely-re-evaluate-your-arguments.